Implement initial support for "lxd init --preseed"

freeekanayaka opened the pull request:
This is the first pass to implement #2834.

It consists of 4 commits that incrementally add support for preseeding the various entities we want.

What it does:

Support running "lxd init --preseed" and feeding it with YAML via standard input to create storage pools, network devices, profiles and set server parameters. For example:

```
  # cat <<EOF | lxd init --preseed
config:
  core.https_address: 127.0.0.1:9999
  images.auto_update_interval: 15
pools:
- name: data
  driver: dir
  config:
    source: ""
networks:
- name: lxd-my-bridge
  type: bridge
  config:
    ipv4.address: none
    ipv6.address: none
profiles:
- name: default
  devices:
    root:
      path: /
      pool: data
      type: disk
- name: test-profile
  description: "Test profile"
  config:
    limits.memory: 2GB
  devices:
    test0:
      name: test0
      nictype: bridged
      parent: lxd-my-bridge
      type: nic
EOF
```
What it does not (yet) do:

- Add more extensive documentation for the feature. I'd like to first get feedback that this implementation is fine (name of the command line switch, format of the YAML, etc) and then write the docs. I'm also not entirely sure where they should go (in a new preseed.md file under doc/ ?).

- Support shortcuts to avoid having to explicitly attach storage pools or network devices to the default profile. This is possibility is mentioned in the issue, but it's hand-wavy for now, we should define how it works first, and it can be considered a follow-up improvement.

- Support passing an argument name to the --preseed command line switch. If you pass "-" that would mean read the YAML from standard in (i.e. the same as you pass nothing), if you pass something other than "-" it will be interpreted as the path to a file to read, containing the YAML. This can be done as follow-up to.

- Cleanup the logic in main_init.go a bit more, so the various auto/interactive/preseed code paths share more code, and easier to understand maintain and extend. I have a few commits for this that I'll submit in a separate PR when this one lands.

freeekanayaka commented:
jenkins: test this please

brauner commented on lxd/main_init.go:
```diff
@@ -494,8 +523,46 @@ they otherwise would.
+func (cmd *CmdInit) runPreseed(client lxd.ContainerServer) error {
+	data := &cmdInitData{}
+
+	if err := cmd.Context.InputYAML(data); err != nil {
```
We don't use the `if err := foo(); err != nil {` syntax anymore and try to eliminate it wherever we see it. Please use:

```
err := foo()
if err != nil {
        return something
}
```

:)

brauner commented on lxd/main_init.go:
```diff
@@ -564,10 +631,18 @@ func (cmd *CmdInit) setProfileConfigItem(c lxd.ContainerServer, profileName stri
 }
 
+// Defines the schema for all possible configuration knobs supported by the
+// lxd init command, either directly feeded via --preseed or populated by
```
nit: `s/feeded/fed/`

brauner commented on lxd/main_init.go:
```diff
@@ -40,6 +41,40 @@ type CmdInit struct {
+
+// Run the logic for auto or interactive mode.
+//
+// IMPLEMENTATION NOTE FOR REVIEWERS (free): this logic is going to be
```
We stay away from adding our names to comments as this gets easily confusing. When we're interested we query `git blame`. :)

freeekanayaka commented:
Christian Brauner <notifications@github.com> writes:

> brauner commented on this pull request.
>
>
>
>> @@ -494,8 +523,46 @@ they otherwise would.
>  			return err
>  		}
>  	}
> +	return nil
> +}
> +
> +// Run the logic for preseed mode
> +func (cmd *CmdInit) runPreseed(client lxd.ContainerServer) error {
> +	data := &cmdInitData{}
> +
> +	if err := cmd.Context.InputYAML(data); err != nil {
>
> We don't use the `if err := foo(); err != nil {` syntax anymore and try to eliminate it wherever we see it. Please use:
>
> ```
> err := foo()
> if err != nil {
>         return something
> }
> ```

Fixed, here and in the other places in this file.

freeekanayaka commented:
Christian Brauner <notifications@github.com> writes:

> brauner commented on this pull request.
>
>
>
>> @@ -564,10 +631,18 @@ func (cmd *CmdInit) setProfileConfigItem(c lxd.ContainerServer, profileName stri
>  	return nil
>  }
>  
> +// Defines the schema for all possible configuration knobs supported by the
> +// lxd init command, either directly feeded via --preseed or populated by
>
> nit: `s/feeded/fed/`

Fixed.

freeekanayaka commented:
Christian Brauner <notifications@github.com> writes:

> brauner commented on this pull request.
>
>
>
>> +	if cmd.Args.Preseed {
> +		err = cmd.runPreseed(client)
> +	} else {
> +		err = cmd.runAutoOrInteractive(client)
> +	}
> +
> +	if err == nil {
> +		cmd.Context.Output("LXD has been successfully configured.\n")
> +	}
> +
> +	return err
> +}
> +
> +// Run the logic for auto or interactive mode.
> +//
> +// IMPLEMENTATION NOTE FOR REVIEWERS (free): this logic is going to be
>
> We stay away from adding our names to comments as this gets easily
> confusing. When we're interested we query `git blame`. :)

Fixed. I replaced it with an "XXX" comment. It should get dropped in a
few branches.

albertodonato commented on lxd/main.go:
```diff
@@ -15,6 +15,7 @@ import (
 
 // Global arguments
 var argAuto = gnuflag.Bool("auto", false, "")
+var argPreseed = gnuflag.Bool("preseed", false, "")
```
It'd be nice to make the option accept a filename, or - for stdin

freeekanayaka replied:
Sounds reasonable to me. We could make "-" the default value (so you can still use "--preseed" without indicating anything else), hope gnuflag supports that.

If there's agreement, I'll change this, but in a follow-up branch.

stgraber replied:
I don't believe gnuflag supports that.

I suspect the easiest will be to always have it take a filename and just support "-" as meaning /dev/stdin.

freeekanayaka replied:
@stgraber do you want to change the behavior in this branch then? Because if we plan to make the filename mandatory , then we can't "formally" do that after merging this branch since it would add a backward incompatibility. If you prefer the semantics suggested by @albertodonato , plus making the filename mandatory, I'd rather implement it in this branch.

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

brauner commented:
Not directly related to your branch but I wonder what https://github.com/freeekanayaka/lxd/blob/14c570537081ea8df2962c39700b655209da4f98/lxd/main_init.go#L344 has got to do in the storage setup part and why it is unset there?

brauner commented on lxd/main_init.go:
```diff
@@ -494,8 +510,148 @@ they otherwise would.
+}
+
+// Apply the configuration specified in the given init data.
+func (cmd *CmdInit) run(client lxd.ContainerServer, data *cmdInitData) error {
```
Ideally this function would be transactional but since that not possible we should at least try to revert to the previous state as best as we can. It can simply be similar to what we do in the `Update()` method in `lxd/container_lxc.go`:
- store the previous state of things (changes, newly created entities and so on)
- define a boolean `tryRevert` or something and then use a closure or a bunch of defers that only fire when `tryRevert` is true and do nothing when `tryRevert` is set to false at the end of the function.

freeekanayaka replied:
Hm if we can't revert completely, I'm not sure to see the value of partial rollback vs clearly stating "You asked me to do X, Y, and Z, I got to X successfully but then Y failed, so I'm giving up".

I assume one change that can't be easily reverted is storage pool creation, which I believe involves some creation of external assets such as zfs/btrs/lvm volumes. What would reverting look like in that case? Delete the state from our db but keep the external assets?

Also, the current behavior of lxd init is not transactional as well (it can fail midway).

Happy to get more input on this, if there's a reasonable and not surprising way to get transactional-ish behavior.

freeekanayaka replied:
On a second thought, I actually realize that we must have already APIs to fully revert things like storage or network bridge creation. So why do you say above it's not possible to have this function fully transactional? Because the "revert" logic could itself fail at some random point? At least in principle.

brauner replied:
You can fully delete a storage pool you just created. The corresponding client function is in `client.go` or at the appropriate place for the new client under `client/*`.

It's not transactional in the sense of "create this set of things (network, pools, etc.) and either succeed or leave the old state". There's a bunch of loops in the functions you wrote and each iteration creates or updates a resource. By design that's not transactional. The revert logic is the closest we get to restoring the old state. If you see a better way though, go ahead.

freeekanayaka replied:
I've now implemented rollback logic in the last commit:

https://github.com/lxc/lxd/pull/3276/commits/a236cec30c286bb22156d9339cd2ab75a306636c

brauner replied:
lgtm

freeekanayaka commented:
Christian Brauner <notifications@github.com> writes:

> Not directly related to your branch but I wonder what
> https://github.com/freeekanayaka/lxd/blob/14c570537081ea8df2962c39700b655209da4f98/lxd/main_init.go#L344
> has got to do in the storage setup part and why it is unset there?

I wondered that too and I found no explanation. Trying to dig the
history with blame didn't help much, but it *seems* that this reset loop
was kind of accidentally brought into the "if storageSetup" block,
without a real intent. That's just speculation tho, so in doubt I kept
it.

freeekanayaka commented on lxd/main_init.go:
```diff
@@ -428,39 +457,22 @@ they otherwise would.
 	}
 
+	if imagesAutoUpdate {
+		data.Config["images.auto_update_interval"] = ""
```
Yes, it's intentional. I started assuming that we wanted this kind of behavior for *both* interactive/auto AND preseed. IOW given that the logic you mentioned has been moved further down, it will still apply to interactive/auto, but it will also apply to preseed. If my assumption is wrong and we want preseed to be more "powerful" and blindly overwrite whatever value it finds, I'll change this (and given the conversation we had a couple of days back about allow overwriting entities when possible, it's indeed probably better to change this part of the PR).

freeekanayaka replied:
BTW, note that there are now unit tests asserting this behavior (e.g. interactive "lxd init" not overwriting a non-zero value different than the default, which is 6).:

https://github.com/lxc/lxd/pull/3276/files#diff-5fa5e50f1b161107bdbf9f02d7436458R128

freeekanayaka replied:
What's your final take? Should I keep this logic/checks only for interactive and auto, and let the plain preseed mode indicate any value? (any value that will then be accepted by the API of course).

freeekanayaka replied:
I went ahead and made the change, since it seems reasonable. Now essentially the old logic of interactive/auto is back to where it was, and the low-level code (used also by preseed) will allow anything.

freeekanayaka commented:
jenkins: test this please

freeekanayaka commented:
jenkins: test this please

stgraber commented on test/suites/init_preseed.sh:
```diff
@@ -0,0 +1,60 @@
+config:
+  core.https_address: 127.0.0.1:9999
+  images.auto_update_interval: 15
+pools:
```
Should be renamed to storage_pools

freeekanayaka replied:
Done.

However I think "storage-pools" might be a tad more appropriate than "storage_pools", since the format and keys of this YAML closely matches the RESTful API, and the endpoint for storage pools is called "/storage-pools" there, not "storage_pools".

stgraber commented on test/suites/init_preseed.sh:
```diff
@@ -0,0 +1,60 @@
+- name: data
+  driver: dir
+  config:
+    source: ""
```
Not the best example since in such case you just wouldn't pass "config" at all.

brauner replied:
Yeah, would be cool if you could remove the `config` section here.

freeekanayaka replied:
I've populated the config in case the zfs backend is used. It gives slightly broader coverage to the feature.

stgraber commented:
So I'm going to trust @brauner on the code as he's been looking at this branch on and off.

What we need there though is documentation as the only example of input I could find was in this PR's description and in the test you wrote.

We still need to do some more reorg of our doc but for now, please create a doc/preseed.md file which explains this feature, what kind of input is valid and the behavior in general.

I'm especially looking for:
 - How to use this to configure a blank LXD?
 - What happens if I run this against a configured LXD (are extra objects removed? what happens if something already exists? ...)

freeekanayaka commented:
@stgraber I've added a doc/preseed.md file with the information you pointed, and addressed your other comments.

brauner commented on lxd/main_init.go:
```diff
@@ -555,8 +542,29 @@ func (cmd *CmdInit) initConfig(client lxd.ContainerServer, config map[string]int
+	// value, and we're being requested to change it to a value
+	// other than zero, we don't want to overwrite it, so we'just
+	// delete it from the desired config.
+	if curVal, ok := server.Config["images.auto_update_interval"]; ok && curVal != "0" {
```
I think I might have mentioned this before and you don't need to change this now but for the future:

```
curVal, ok := server.Config["images.auto_update_interval"]
if ok && curVal != "0" {
        // something
}
```

to keep the clode nice an uncluttered. :)

freeekanayaka replied:
Ah okay :)

brauner commented on lxd/main_init.go:
```diff
@@ -555,8 +542,29 @@ func (cmd *CmdInit) initConfig(client lxd.ContainerServer, config map[string]int
+		// The underlying code expects all values to be string, even if when
+		// using preseed the yaml.v2 package unmarshals them as integers.
+		if number, ok := value.(int); ok {
+			value = strconv.Itoa(number)
```
I assume that there are no values that can be floats.

freeekanayaka replied:
Not that I see in daemon_config.go

brauner commented on lxd/main_init.go:
```diff
@@ -573,15 +583,62 @@ func (cmd *CmdInit) initConfig(client lxd.ContainerServer, config map[string]int
+	for _, pool := range pools {
+		_, _, err := client.GetStoragePool(pool.Name)
+		if err == nil {
+			err = client.UpdateStoragePool(pool.Name, api.StoragePoolPut{
```
The commit message mentions that updating storage pools will always fail? :)

freeekanayaka replied:
Yeah, I think for the most part this API call will fail, but I guess there are small things like the new description field that one can update without the API call failing. Anyway, this code ensures that the call is tried, it will be up the the API implementation to decide if it can succeed or not (and that's effectively the real meaning of the mention in the commit message).

brauner commented on lxd/main_init.go:
```diff
@@ -825,7 +825,7 @@ func (cmd *CmdInit) setProfileConfigItem(c lxd.ContainerServer, profileName stri
 type cmdInitData struct {
 	api.ServerPut `yaml:",inline"`
-	Pools         []api.StoragePoolsPost
+	Pools         []api.StoragePoolsPost `yaml:"storage_pools"`
```
Why only for `StoragePoolsPost` and not `NetworksPost` and others?

freeekanayaka replied:
Because if the name of structure field is called "Foo" the go yaml package will map it to "foo" when marshalling and unmarshalling. So "Networks" gets mapped to "networks" by default without an explicit annotation.

I can surely add annotations everywhere for extract clarity, if you prefer.

freeekanayaka commented:
jenkins: test this please

conquistadorConsulting commented:
I have lxd version 2.0.11 and I am getting this error: "error: flag provided but not defined: --preseed". Why?

stgraber commented:
Because 2.0.11 is lower than 2.14. We don't backport new features to the LTS branch (2.0.x), only bug and security fixes.

conquistadorConsulting commented:
How do I upgrade lxd 2.0.11 to lxd 2.14? I executed "sudo apt-get update" and everything is current?

stgraber commented:
@conquistadorConsulting apt install -t xenial-backports lxd lxd-client

That will upgrade you from 2.0.11 to 2.21. Do note that 2.21 isn't a LTS release for LXD, so you may get frequent (monthly) upgrades to your LXD daemon, introducing new features and changed behaviors.
